jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/PushApplicationEndpoint.java

```diff
@@ -70,27 +71,34 @@
      *
      * @statuscode 201 The PushApplication Variant created successfully
      * @statuscode 400 The format of the client request was incorrect
+     * @statuscode 409 The PushApplication already exists
      */
     @POST
     @Consumes(MediaType.APPLICATION_JSON)
     @Produces(MediaType.APPLICATION_JSON)
     @ReturnType("org.jboss.aerogear.unifiedpush.api.PushApplication")
     public Response registerPushApplication(PushApplication pushApp) {
 
-         // some validation
         try {
             validateModelClass(pushApp);
         } catch (ConstraintViolationException cve) {
+            return createBadRequestResponse(cve.getConstraintViolations()).build();
+        }
 
-            // Build and return the 400 (Bad Request) response
-            ResponseBuilder builder = createBadRequestResponse(cve.getConstraintViolations());
-
-            return builder.build();
+        try {
+            pushAppService.addPushApplication(pushApp);
+        } catch (IllegalArgumentException e) {
```

Comment:
This works when the application with the same ID is already committed.
I'm not sure it will work if two different clients try to create two new apps with the same ID at the same time, since the conflict will be notified only when the transaction will be committed.
@matzew What are your thoughts?

Reply:
You are right. The only to way to be certain of this is to call `.flush()` right after creating the _add_ transaction. This is whatsoever a bit ugly IMO.

Reply:
Nope. Even if you flush you won't get the error if the other transaction is not committed.
The flush just applies the changes to the transaction.
The only place you will be sure to get the error is where the transaction is committed.